### PR TITLE
Feature nxp 24251 web ui publishing

### PR DESCRIPTION
**Pull request by guirenard**

### CODE REVIEW

File: elements/nuxeo-app.html [hunk 1]
```diff
@@ -1161,8 +1161,9 @@
 
       _computeEnrichers: function() {
         return (Nuxeo.UI && Nuxeo.UI.config && Nuxeo.UI.config.enrichers && Nuxeo.UI.config.enrichers.join(',')) ||
-        'hasContent, firstAccessibleAncestor, permissions, breadcrumb, preview, favorites, subscribedNotifications,' +
-        ' thumbnail, renditions, pendingTasks, runnableWorkflows, runningWorkflows, collections, audit, subtypes, tags';
+        'hasContent, firstAccessibleAncestor, permissions, breadcrumb, preview, favorites, subscribedNotifications' +
```
Comment (nelsonsilva) [inline elements/nuxeo-app.html]: nit: this is getting hard to follow, perhaps we should use `[...].join(',')` to have one per line

File: elements/nuxeo-document-info-bar/nuxeo-document-info-bar.html [hunk 2]
```diff
@@ -158,6 +159,11 @@
       </div>
     </template>
 
+    <!-- proxy/publication -->
+    <template is="dom-if" if="[[isProxy(document)]]">
```
Comment (nelsonsilva) [inline elements/nuxeo-document-info-bar/nuxeo-document-info-bar.html]: All proxies are publications?

File: elements/nuxeo-publication/nuxeo-document-publications.html [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+      },
+
+      _unpublish: function(e) {
+        if (e && e.target) {
+          if (!confirm(this.i18n('publication.unpublish.confirm'))) {
+            return;
+          }
+          var doc = e.target.parentNode.item;
+          this.$.unpublishOp.input = doc;
+          this.$.unpublishOp.execute().then(function() {
```
Comment (nelsonsilva) [inline elements/nuxeo-publication/nuxeo-document-publications.html]: Q: wdyt about having nuxeo-operation take optional input and parameters to make this cleaner ?

File: elements/nuxeo-publication/nuxeo-internal-publish.html [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+        selectedRendition: {
+          type: String,
+          value: 'none'
+        },
+
+        keepAlreadyPublished: Boolean
+      },
+
+      _computeRenditionOptions: function() {
```
Comment (nelsonsilva) [inline elements/nuxeo-publication/nuxeo-internal-publish.html]: where is this used ?

File: elements/nuxeo-publication/nuxeo-internal-publish.html [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        }
+      },
+
+      _publish: function() {
+        this.$.op.params = {
+          'target': this.publishSpace.uid,
+          'override': this.override
+        }
+        if (this.selectedRendition) {
+          if (this.selectedRendition === 'default') {
```
Comment (nelsonsilva) [inline elements/nuxeo-publication/nuxeo-internal-publish.html]: Why does the op not work with `renditionName = 'default'` ?

File: elements/nuxeo-publication/nuxeo-internal-publish.html [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            this.$.op.params.defaultRendition = true;
+          } else if (this.selectedRendition !== 'none') {
+            this.$.op.params.renditionName = this.selectedRendition;
+          }
+        }
+        this.$.op.input = this.document.uid;
+        this.$.op.execute().then(function(publishedDoc) {
+          this.fire('notify', {'message': this.i18n('publication.internal.publish.success', publishedDoc.path)});
+          this.fire('document-updated');
+          this.fire('nx-publish-success');
```
Comment (nelsonsilva) [inline elements/nuxeo-publication/nuxeo-internal-publish.html]: Are we using the `nx-` prefix for events ?

File: elements/nuxeo-publication/nuxeo-publication-info-bar.html [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      iron-icon {
+        margin: 0 .5em;
+        width: 1.5em;
+      }
+    </style>
+
+    <nuxeo-document id="srcDoc" loading="{{loadingSrc}}">
+    </nuxeo-document>
+
+    <div id="publicationInfoBar">
```
Comment (nelsonsilva) [inline elements/nuxeo-publication/nuxeo-publication-info-bar.html]: Do we need this wrapping div? Can't we just style `:host`?

File: elements/nuxeo-publication/nuxeo-publication-info-bar.html [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      <div class="layout horizontal center" hidden$="[[loadingSrc]]">
```
Comment (nelsonsilva) [inline elements/nuxeo-publication/nuxeo-publication-info-bar.html]: Seems this div isn't really doing anything as you've applied the mixins to the parent

File: elements/nuxeo-publication/nuxeo-unpublish-button.html [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                            dialog-confirm="[[i18n('label.yes')]]"
+                            on-confirm="_unpublish">
+        <span class="label">[[i18n('publication.unpublish')]]</span>
+      </nuxeo-confirm-button>
+    </template>
+
+  </template>
+
+  <script>
+    Polymer({
```
Comment (nelsonsilva) [inline elements/nuxeo-publication/nuxeo-unpublish-button.html]: Need to align on the more actions stuff ?

File: elements/nuxeo-web-ui-bundle.html [hunk 10]
```diff
@@ -370,7 +376,36 @@
   </template>
 </nuxeo-slot-content>
 
-<nuxeo-slot-content name="documentTrashItem" slot="DOCUMENT_VIEWS_ITEMS" order="40">
+<nuxeo-slot-content name="documentPublicationItem" slot="DOCUMENT_VIEWS_ITEMS" order="40">
+  <template>
+    <nuxeo-filter document="[[document]]" expression="!document.isProxy
```
Comment (nelsonsilva) [inline elements/nuxeo-web-ui-bundle.html]: 😱 can we at least add a helper in nuxeo-filter to make this a shorter and cleaner ?

File: elements/nuxeo-web-ui-bundle.html [hunk 11]
```diff
@@ -637,3 +672,23 @@
...
   </template>
 </nuxeo-slot-content>
+
+<!-- PUBLISHING -->
+<nuxeo-slot-content name="internalPublishItem" slot="PUBLISH_ITEMS" order="10">
+  <template>
+    <paper-tab name="internal">[[i18n('publication.internal.title')]]</paper-tab>
+  </template>
+</nuxeo-slot-content>
+<!--
```
Comment (nelsonsilva) [inline elements/nuxeo-web-ui-bundle.html]: Do we want keep this?

File: elements/nuxeo-publication/nuxeo-document-publications.html [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+            this.fire('notify', {'message': this.i18n('publication.unpublish.success')});
+            this._fetchPublications();
+          }.bind(this)).catch(function() {
+            this.fire('notify', {'message': this.i18n('publication.unpublish.error')});
+          }.bind(this));
+        }
+      },
+
+      _republish: function() {
+        alert('Not implemented yet!');
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-document-publications.html]: yes, coming this sprint

File: elements/nuxeo-document-info-bar/nuxeo-document-info-bar.html (hunk 2 above)
Comment (guirenard) [reply elements/nuxeo-document-info-bar/nuxeo-document-info-bar.html]: good question, maybe @efge can help answering?

File: elements/nuxeo-publication/nuxeo-publication-info-bar.html (hunk 7 above)
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-publication-info-bar.html]: :man_shrugging:

File: elements/nuxeo-document-info/nuxeo-document-info.html [hunk 13]
```diff
@@ -99,9 +114,20 @@
...
+      return this.document && this.hasFacet(this.document, 'Publishable') && this.document.contextParameters &&
+        this.document.contextParameters.publications && this.document.contextParameters.publications.resultsCount > 0;
+    },
+
     _documentChanged: function() {
       this._showProcess = this.document && this.document.contextParameters &&
         this.document.contextParameters.runningWorkflows && this.document.contextParameters.runningWorkflows.length > 0;
+    },
+
+    _urlForPub: function() {
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-document-info/nuxeo-document-info.html]: ok I created https://jira.nuxeo.com/browse/ELEMENTS-789

File: elements/nuxeo-document-info/nuxeo-document-info.html [hunk 14]
```diff
@@ -99,9 +114,20 @@
...
+      if (this.document) {
+        return this.urlFor('browse', this.document.path) + '?p=publication';
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-document-info/nuxeo-document-info.html]: will proceed with https://jira.nuxeo.com/browse/ELEMENTS-789

File: elements/nuxeo-publication/nuxeo-document-publications.html (hunk 3 above)
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-document-publications.html]: You mean between Unpublish and UnpublishAll ? We could, but I feel like the signature will be very fuzzy between required input/param  for document and publication depending on what you want to do.

File: elements/nuxeo-publication/nuxeo-internal-publish.html (hunk 4 above)
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-internal-publish.html]: :+1: cleaning coming

File: elements/nuxeo-publication/nuxeo-internal-publish.html (hunk 5 above)
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-internal-publish.html]: This is related to https://github.com/nuxeo/nuxeo/pull/2414#discussion_r223336830

File: elements/nuxeo-publication/nuxeo-internal-publish.html (hunk 6 above)
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-internal-publish.html]: not always, but I guess we should :|

File: elements/nuxeo-publication/nuxeo-publication-info-bar.html (hunk 7 above)
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-publication-info-bar.html]: :+1:

File: elements/nuxeo-publication/nuxeo-publication-info-bar.html (hunk 8 above)
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-publication-info-bar.html]: :+1:

File: elements/nuxeo-publication/nuxeo-publication-info-bar.html [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        <iron-icon icon="icons:info"></iron-icon>
+        <div>
+          <a class="" href$="[[_srcUrl(_src)]]" disabled$="[[_srcDeleted]]">
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-publication-info-bar.html]: :+1:

File: elements/nuxeo-publication/nuxeo-publication-info-bar.html [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+        /**
+         * Input document.
+         */
+        document: Object,
+        _redirectDoc: Object,
+        _src: Object,
+        _srcDeleted: Boolean
+      },
+      observers: [
+        '_updateSrc(document)',
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-publication-info-bar.html]: :+1:

File: elements/nuxeo-publication/nuxeo-publication-info-bar.html [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        this._srcDeleted = false;
+        if (this.document && this.document.isProxy) {
+          this.$.srcDoc.docId = this.document.versionableId;
+          this.$.srcDoc.get().then(function(src) {
+            this._src = src;
+            this._redirectDoc =
+              (this.document.contextParameters && this.document.contextParameters.firstAccessibleAncestor) || src;
+          }.bind(this)).catch(function(err) {
+            if (err.status === 404) {
+              this._srcDeleted = true;
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-publication-info-bar.html]: The thing is while we load the source, we may show wrong label for a short while if we don't have such property

File: elements/nuxeo-publication/nuxeo-publish-button.html [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+      :host {
+        --paper-tabs-selection-bar-color: var(--nuxeo-primary-color);
+        display: block;
+      }
+
+      nuxeo-dialog {
+        min-width: 915px;
+      }
+
+      @media (max-width: 1024px) {
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-publish-button.html]: I'll clean that up. Not sure this is needed anw. For a mixin, I guess we should have a dedicated ticket to review all dialogs

File: elements/nuxeo-publication/nuxeo-unpublish-button.html (hunk 9 above)
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-unpublish-button.html]: Supposedly done, what do you think it is missing?

File: elements/nuxeo-web-ui-bundle.html (hunk 11 above)
Comment (guirenard) [reply elements/nuxeo-web-ui-bundle.html]: no :+1:

File: elements/nuxeo-web-ui-bundle.html (hunk 10 above)
Comment (guirenard) [reply elements/nuxeo-web-ui-bundle.html]: will do

File: elements/nuxeo-app.html (hunk 1 above)
Comment (guirenard) [reply elements/nuxeo-app.html]: :+1:

File: elements/nuxeo-document-info-bar/nuxeo-document-info-bar.html (hunk 2 above)
Comment (guirenard) [reply elements/nuxeo-document-info-bar/nuxeo-document-info-bar.html]: We have indeed the live proxies which are multi-place documents, I will try to work it out, it seems they don't have the `Immutable` facet

File: elements/nuxeo-document-versions/nuxeo-document-versions.html [hunk 19]
```diff
@@ -41,6 +41,7 @@
     <style>
       :host {
         display: inline-block;
+        min-width: 60px;
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-document-versions/nuxeo-document-versions.html]: yes, it's because I am reusing this widget in my publish dialog and depending on window size, it was missing some width

File: elements/nuxeo-publication/nuxeo-document-publications.html [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+      <div class="resultActions">
+        <paper-button class="uppercase" on-tap="_unpublishAll" disabled$="[[!hasPublications]]">[[i18n('publication.unpublishAll')]]</paper-button>
+        <paper-button class="uppercase primary" on-tap="_republishAll" disabled$="[[!hasPublications]]">[[i18n('publication.republishAll')]]</paper-button>
+      </div>
+      <nuxeo-data-table id="table" nx-provider="provider" class="results" items="{{publishedDocs}}"
+                        empty-label="[[i18n('publication.noPublications')]]">
+        <nuxeo-data-table-column name="[[i18n('documentContentView.datatable.header.title')]]" flex="200">
+          <template>
+            <nuxeo-document-thumbnail document="[[item]]"></nuxeo-document-thumbnail>
+            <a class="path ellipsis" href$="[[urlFor('browse', item.path)]]" on-tap="_navigate">[[item.path]]</a>
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-document-publications.html]: my bad, I meant `on-row-clicked="_navigate"` on the data-table to be consistent with other tables in the application.

File: elements/nuxeo-publication/nuxeo-internal-publish.html [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+    </style>
+
+    <nuxeo-operation id="op" op="Document.PublishToSection">
+    </nuxeo-operation>
+
+    <nuxeo-document id="srcDoc">
+    </nuxeo-document>
+
+    <div class="container">
+      <paper-dialog-scrollable>
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-internal-publish.html]: Ok, but this goes beyond my skills in css. Just looking at https://github.com/PolymerElements/paper-dialog-scrollable/blob/2.x/paper-dialog-scrollable.html makes me think it is much more complicated than applying mixins

File: elements/nuxeo-publication/nuxeo-internal-publish.html [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+          }
+        }
+        this.$.op.input = this.document.uid;
+        this.$.op.execute().then(function(publishedDoc) {
+          this.fire('notify', {'message': this.i18n('publication.internal.publish.success', publishedDoc.path)});
+          this.fire('document-updated');
+          this.fire('nx-publish-success');
+        }.bind(this)).catch(function(err) {
+          this.fire('notify', {'message': this.i18n('publication.internal.publish.error', err)});
+          this.fire('nx-publish-error');
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-internal-publish.html]: :+1: or nothing, just notify and rethrow the execption

File: elements/nuxeo-publication/nuxeo-publish-button.html [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+        _label: {
+          type: String,
+          computed: '_computeLabel(i18n)'
+        }
+      },
+
+      listeners: {
+        'nx-publish-cancel': '_close',
+        'nx-publish-success': '_close',
+        'nx-publish-error': '_close'
```
[the remark this answers is not in the document]
Comment (guirenard) [reply elements/nuxeo-publication/nuxeo-publish-button.html]: not really :+1:

File: elements/nuxeo-publication/nuxeo-internal-publish.html (hunk 21 above)
[the remark this answers is not in the document]
Comment (nelsonsilva) [reply elements/nuxeo-publication/nuxeo-internal-publish.html]: Yes, sorry about that, I started writing then looked into `paper-dialog-scrollable` and reached the same conclusion. just kept the rambling to share the pain :P
